File: clients/typescript/src/debugging/debug-state.ts
import { Stream } from "../stream";

/**
 * App state for debugging
 */
export class DebugState {
    /**
     * Currently pressed keys
     */
    pressedKeys: Array<string>

    constructor(pressedKeys: Array<string>) {
        this.pressedKeys = pressedKeys;
    }

    /**
     * Read DebugState from input stream
     */
    static async readFrom(stream: Stream): Promise<DebugState> {
        let pressedKeys;
        pressedKeys = [];
        for (let pressedKeysCount = await stream.readInt(); pressedKeysCount > 0; pressedKeysCount--) {
            let pressedKeysElement;
            pressedKeysElement = await stream.readString();
            pressedKeys.push(pressedKeysElement);
        }
        return new DebugState(pressedKeys)
    }

    /**
     * Write DebugState to output stream
     */
    async writeTo(stream: Stream) {
        let pressedKeys = this.pressedKeys;
        await stream.writeInt(pressedKeys.length);
        for (let pressedKeysElement of pressedKeys) {
            await stream.writeString(pressedKeysElement);
        }
    }
}

File: clients/java/src/main/java/komus24/debugging/DebugState.java
package komus24.debugging;

import komus24.util.StreamUtil;

/**
 * App state for debugging
 */
public class DebugState {
    /**
     * Currently pressed keys
     */
    private String[] pressedKeys;

    /**
     * Currently pressed keys
     */
    public String[] getPressedKeys() {
        return pressedKeys;
    }

    /**
     * Currently pressed keys
     */
    public void setPressedKeys(String[] value) {
        this.pressedKeys = value;
    }

    public DebugState(String[] pressedKeys) {
        this.pressedKeys = pressedKeys;
    }

    /**
     * Read DebugState from input stream
     */
    public static DebugState readFrom(java.io.InputStream stream) throws java.io.IOException {
        String[] pressedKeys;
        pressedKeys = new String[StreamUtil.readInt(stream)];
        for (int pressedKeysIndex = 0; pressedKeysIndex < pressedKeys.length; pressedKeysIndex++) {
            String pressedKeysElement;
            pressedKeysElement = StreamUtil.readString(stream);
            pressedKeys[pressedKeysIndex] = pressedKeysElement;
        }
        return new DebugState(pressedKeys);
    }

    /**
     * Write DebugState to output stream
     */
    public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
        StreamUtil.writeInt(stream, pressedKeys.length);
        for (String pressedKeysElement : pressedKeys) {
            StreamUtil.writeString(stream, pressedKeysElement);
        }
    }

    /**
     * Get string representation of DebugState
     */
    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder("DebugState { ");
        stringBuilder.append("pressedKeys: ");
        stringBuilder.append("[ ");
        for (int pressedKeysIndex = 0; pressedKeysIndex < pressedKeys.length; pressedKeysIndex++) {
            if (pressedKeysIndex != 0) {
                stringBuilder.append(", ");
            }
            String pressedKeysElement = pressedKeys[pressedKeysIndex];
            stringBuilder.append('"' + pressedKeysElement + '"');
        }
        stringBuilder.append(" ]");
        stringBuilder.append(" }");
        return stringBuilder.toString();
    }
}